src/utils/PluginManager.ts

```diff
@@ -0,0 +1,121 @@
+          // this is a plugin that can be acquired from npm.
+          // it might already be installed, so check for that first.
+          const existingInstallationPath = path.join(pluginBasePath, 'node_modules', plugin.name);
+          let shouldPerformInstallation = true;
+          if (fs.existsSync(path.join(existingInstallationPath, 'package.json'))) {
+            // check the version number in package.json
+            const packageContents = fs.readJsonSync(
+              path.join(existingInstallationPath, 'package.json')
+            );
+            if (packageContents.version === plugin.version) {
+              shouldPerformInstallation = false;
+            }
+          }
+          if (shouldPerformInstallation) {
+            await PluginManager.installFromNpm(pluginBasePath, `${plugin.name}@${plugin.version}`);
+          }
+          PluginManager.loadFromFilesystem([existingInstallationPath], plugin.name, plugin.args);
+        } else {
+          // this is a plugin managed on the local filesystem.
+          // it may be at the base path, or in the node_modules directory
+          PluginManager.loadFromFilesystem(
+            [
+              path.join(pluginBasePath, plugin.name),
+              path.join(pluginBasePath, 'node_modules', plugin.name)
+            ],
+            plugin.name,
+            plugin.args
+          );
+        }
+      } catch (err) {
+        logger.error(`Couldn't load plugin ${plugin.name}: ${err.message}`);
+      }
+    }
+  }
+
+  // a small test for name safety, based on typical naming patterns for npm packages
+  static isNameNpmSafe(name: string): boolean {
+    return /^[\w\d@/._-]+$/.test(name);
+  }
+
+  static async installFromNpm(installPath: string, installTarget: string) {
+    if (PluginManager.isNameNpmSafe(installTarget)) {
+      fs.ensureDirSync(installPath);
+      const installationResult = await util.promisify(execFile)(
+        'npm',
+        ['install', '--prefix', '.', installTarget],
+        {
+          cwd: installPath,
+          shell: true
+        }
+      );
+      if (installationResult.stderr.length > 0) {
+        throw new Error(installationResult.stderr);
+      } else {
+        logger.info(`Installed plugin ${installTarget}`);
+      }
+    } else {
+      throw new QuestionableNpmNameError();
+    }
+  }
+
+  static loadFromFilesystem(basePaths: string[], name: string, pluginArgs: any[] = []) {
+    for (const basePath of basePaths) {
+      const pluginPath = path.resolve(basePath);
+      if (fs.existsSync(pluginPath) && fs.statSync(pluginPath).isDirectory()) {
+        // eslint-disable-next-line @typescript-eslint/no-var-requires
+        const pluginModule = require(pluginPath);
+        // does the module actually have an initialize function?
+        if (pluginModule.initialize != null && typeof pluginModule.initialize === 'function') {
+          pluginModule.initialize(PluginManager.registerHook, ...pluginArgs);
```

Comment:
I wonder if the `initialize` function should be asynchronous, just in case the plugin needs to do anything fancy?  I tested w/ an async initialize function and since this does not `await` it, it never got resolved.